### PR TITLE
[flang][OpenMP] Extend `do concurrent` mapping to device.

### PR DESCRIPTION
**Pull request by ergawy**
For simple loops, we can now choose to map `do concurrent` to either the host (i.e. `omp parallel do`) or the device (i.e. `omp target teams distribute parallel do`).

In order to use this from `flang-new`, you can pass: `-fdo-concurrent-parallel=[none|host|device]`.

### CODE REVIEW

**Review by mjklemm — APPROVED** [review]
LGTM

**Comment by ergawy** [comment]
Thanks @mjklemm for the review.

Had some difficulty merging with ATD, but now merged and should be ready for review again in case somebody else in addition to Michael has time for review.

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:124]: Nit: The general consensus is to avoid hardcoding some arbitrary `llvm::SmallVector` size in the interface of a class (in this case it's the default implicitly-calculated one). I'd generally tend towards doing something like `void collectLoopLiveIns(fir::DoLoopOp doLoop, llvm::SmallVectorImpl<mlir::Value> &liveIns) const`, but I suppose if you really wanted to return the vector and rely on compiler optimizations to avoid copies you could do `  llvm::SmallVectorImpl<mlir::Value> collectLoopLiveIns(fir::DoLoopOp doLoop) const` as well.

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:174]:
```suggestion
    auto declareOp =
```

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:205]:
```suggestion
      auto shapeOp = mlir::dyn_cast_if_present<fir::ShapeOp>(
```

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:403]: I think it makes sense to copy the ops related to constants and type casts into the target region (rather than just before `omp.wsloop`, since that means introducing operations between the leaf constructs of a composite, which will be disallowed by loop wrappers), but shouldn't everything be already mapped and present as an entry block argument to `omp.target` by this time, due to it being detected as a live-in value?

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:82]:
```suggestion
    mlir::omp::TargetOp targetOp;
```

**Review by skatrak — COMMENTED** [review]
Thank you Kareem for this work. I seem to be a bit too slow today to actually understand everything that's going on here, but I left you some of my thoughts going through it. In any case, it looks like a good step to get there. I'll probably better leave it to @agozillon to comment on the map-related part of the PR.

One thing to note is that this approach will have to change a bit with the next merge from main, which will hopefully already bring the changes to `omp.wsloop`. The main conflict with your work will be that `omp.wsloop %i... { body }` will have to be replaced by `omp.wsloop { omp.loop_nest %i... { body } }` and no operations will be allowed between the nested `omp.distribute`, `omp.parallel`, `omp.wsloop` and `omp.loop_nest` operations because they would be loop wrappers representing a composite construct. I'm not sure how long it will be before that happens, but I'm hoping for about a week, depending on reviews and so on. If it's not too much of a blocker for you, I'd prefer if you could wait until then before merging this (even if it's ready before), so that any conflicts specific to your PR can be addressed separately, I could help your with that as well.

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (agozillon) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:98]: seems like a bit of the sentence is missing here! Perhaps just the word "set"

File: flang/include/flang/Lower/OpenMP/Utils.h
Comment (agozillon) [inline flang/include/flang/Lower/OpenMP/Utils.h:53]: I'm not against changing this to the more generic mlir::OpBuilder, but I am wondering if we should just keep it the same as most of the other flang utilities to keep it somewhat standardized, but I'll leave that up to you!

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (agozillon) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:222]: You can likely verify the calculation in the DirectiveCommons.h code where the bounds are generated, and if still in doubt fire a message towards one of the OpenACC guys who created it, there's been a bug or two before, but they're generally quite good at clarifying things :-)

But I suppose if the lower bound is 0, and the extent is the max size of a dimension, you'd get the upper bound, but the subtraction seems like extra legwork at a minimum here I think, as you could just take the extent I would imagine.

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (agozillon) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:236]: The map code looks reasonable for the moment! I would just advise that you make sure this pass occurs prior to the OpenMP descriptor pass otherwise the descriptor types (allocatables et al) that are captured won't be expanded appropriately which will result in them not being mapped correctly. I think that should be enough to have these types mapped implicitly.

You probably already have the passes pipelined like this but just wanted to make sure!

Comment (agozillon) [reply flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:236]: This little segment of code will also need to maintain lockstep with any changes to the implicit mapping in the OpenMP lowering as well unfortunately. There is nothing new yet, but there likely will be in the future but I can perhaps update this or keep you informed of any changes I make (or I see being made) :-)

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (agozillon) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:274]: seems reasonable from what I can see, but I would also keep in mind that during the generation of Map <-> BlockArgs that occurs, we also have a little segment of code that clones bounds into the target region where possible (I think genLiveInDeclare may be trying to do something similar but please do correct me if I am wrong), and if it can't I believe it generates a map for the bounds, but @TIFitis is more familiar with this and can perhaps give further input or answer questions you might have on it!

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (agozillon) [reply flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:403]: I think one issue that may arise, is that map operations only take pointer-like values, Index values are not mappable (you can circumvent this by creating the right operations to load it into memory (so convert it to a ref) and pass that instead I would imagine), there was some contention with this when @TIFitis tried to do it originally, so we rolled back to just using pointer-like arguments in map operations.

**Review by agozillon — APPROVED** [review]
LGTM, some minor comments here and there, mostly just some info dumping and two nits, one of which is up to you what you'd like to do :-) 

Of course, please do wait on @skatrak's approval before going forward!

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (ergawy) [reply flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:384]: > but shouldn't everything be already mapped and present as an entry block argument to omp.target by this time, due to it being detected as a live-in value?

For the simple cases I tested so far, this is not the situation since LB, UB, step and constants; so they are not detected as live-ins. Only the induction variable is a live-in.

> I think it makes sense to copy the ops related to constants and type casts into the target region (rather than just before omp.wsloop,

If you don't mind, I will leave that until we merge your work on composite constructs from upstream.

File: flang/include/flang/Lower/OpenMP/Utils.h
Comment (ergawy) [reply flang/include/flang/Lower/OpenMP/Utils.h:81]: Sounds good. I think that will need to happen for both PFT -> OMP & `do concurrent` conversion lowering paths, right? If so, will you add a `TODO` somewhere for that?

**Review by ergawy — COMMENTED** [review]
Thanks @skatrak & @agozillon for the review.

File: flang/lib/Optimizer/Transforms/CMakeLists.txt
Comment (ergawy) [reply flang/lib/Optimizer/Transforms/CMakeLists.txt:26]: That's ugly indeed. I think we should separate everything under `flang/lib/Lower/OpenMP` to a separate target: `OpenMPLower`. But I think we should create a separate an upstream PR for this, WDYT?

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:356]:
```suggestion
    auto cloneBoundOrStepDefChain = [&](mlir::Operation *operation) -> mlir::Operation * {
```

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:334]:
```suggestion
    rewriter.createBlock(&teamsOp.getRegion());
    rewriter.setInsertionPoint(rewriter.create<mlir::omp::TerminatorOp>(loc));
```

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:279]: Nit: It's a commonly used pattern, so I think we don't need to split it into two statements.
```suggestion
    rewriter.setInsertionPoint(rewriter.create<mlir::omp::TerminatorOp>(targetOp.getLoc()));
```

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:253]: Nit: I suppose this is separated from `genTargetOp` because it was copied and adapted from PFT to OpenMP lowering, but it doesn't look to give much of a benefit here, maybe it's better to just inline this into `genTargetOp`.

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:124]:
```suggestion
  /// The first item in the returned list is always the loop's induction
```

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:83]: Nit: No need for explicitly initializing to `nullptr`, as it's the default.
```suggestion
    mlir::omp::TargetOp targetOp;
```

File: flang/lib/Optimizer/Transforms/CMakeLists.txt
Comment (skatrak) [reply flang/lib/Optimizer/Transforms/CMakeLists.txt:27]: We have to do something like that, and I agree that it's something we'd want community comments for. The problem I see at the moment is that the only reason we want this right now is because of this pass, which we can't currently upstream. So, if we create an upstream PR to try to separate this, there won't be any follow-up patches to make use of it.

At this time, I would suggest to just leave it as you wrote it, adding a TODO comment right above this new line so we don't forget this is a workaround. And then, once we find the need to get something upstream that would benefit from sharing some code with lowering, we should try to do it right and remove this in the following merge from main. I hope that makes sense.

**Review by skatrak — COMMENTED** [review]
Sorry for the delay getting back to this, but thank you Kareem for working on previous feedback and adapting the patch to the changes to OpenMP loop representations. I've got a few comments mostly to reorganize things a bit and some nits, but nothing substantial.

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (ergawy) [reply flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:347]: It is a recursive lambda, you have to specify the type.

File: flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp
Comment (skatrak) [reply flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp:420]: Sure, that works as well.

**Pull request merged**
